java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageChannelReader.java

```diff
@@ -18,98 +18,70 @@
 
 package org.apache.arrow.vector.ipc.message;
 
+import java.io.IOException;
 
-import io.netty.buffer.ArrowBuf;
-import org.apache.arrow.flatbuf.Message;
 import org.apache.arrow.memory.BufferAllocator;
 import org.apache.arrow.vector.ipc.ReadChannel;
 
-import java.io.IOException;
-import java.nio.ByteBuffer;
-
 /**
  * Reads a sequence of messages using a ReadChannel.
  */
-public class MessageChannelReader implements MessageReader {
+public class MessageChannelReader extends MessageReader<ArrowBufReadHolder> {
 
-  private ReadChannel in;
+  protected ReadChannel in;
+  protected BufferAllocator allocator;
 
   /**
-   * Construct from an existing ReadChannel.
+   * Construct a MessageReader to read streaming messages from an existing ReadChannel.
    *
    * @param in Channel to read messages from
+   * @param allocator BufferAllocator used to read Message body into an ArrowBuf.
    */
-  public MessageChannelReader(ReadChannel in) {
+  public MessageChannelReader(ReadChannel in, BufferAllocator allocator) {
     this.in = in;
+    this.allocator = allocator;
```

Comment:
This somewhat change the ownership of the arrow memory allocated here from the caller to the reader instance. I feel this maybe an API that needs some more discussion.

Specifically, before the change, when a caller uses this class to read arrow data, the memory ownership is pretty clear - the caller owners the BufferAllocator. However, now the allocator is inside the class and therefore not owned by the caller, does it mean the ownership of the data is now changed? Or does the caller still own the memory, but not the allocator? I found this a little bit hard to reason...

cc @jacques-n WDYT?

Reply:
I don't think this really changes memory ownership.  The `MessageChannelReader` is constructed with a given allocator, it does not create one.  The allocated buffers are owned by the creator of the `MessageChannelReader` instance, which seems like how it should be. And usually the creator of the instance is going to be the same that calls `readNext()`, where the allocation is performed.  I think it's unlikely that one would change allocators from message to message, so this makes things easier.

Reply:
I see. This seems fine.